### PR TITLE
feat(rgpd): nullify ressources when destroy for rgpd reason

### LINKED ISSUE #1376: [RGPD] Processer des webhooks d'inactivité
* Faire en sorte que RDVS envoie un attribut dans le webhook qui veut dire que la ressource est supprimée pour inactivité
* Lorsqu'on reçoit ce webhook, enlever le lien avec la ressource rdvs (en mettant l'id rdv_solidarites à `null`) plutôt que de la supprimer

### LINKED ISSUE #1378: [RGPD] Gérer les ressources non rattachées à RDVS
Il faudrait gérer les états des personnes n'ayant plus de fiche RDVS (`applicants` sans `rdv_solidarites_user_id`): 
* Ne plus afficher de lien vers RDVS
* Créer un usager RDVS et le lier à la personne lorsqu'on fait une action (inviter, trouver un rdv, etc.)
* Voir si on ne devrait pas tout le temps repousser la création de la fiche RDVS à ce moment-là, même lorsque l'on crée la fiche RDVI
* Potentiellement gérer le cas où cet usager est recréé sur RDVS sans que le lien soit fait sur RDVI
* Voir les autres implications métier potentielles

Pour les rdvs c'est plus simple, le fait de juste ne plus afficher le lien devrait suffire.

### PR DESCRIPTION
**Pull request by qblanc**
closes #1376
closes #1378 

Dans cette PR, je traite différemment les ressources supprimées côté rdv-s pour cause de RGPD (inactivité pendant 2 ans côté rdvs-), qui peuvent ne pas devoir être supprimées côté rdv-i. Les ressources concernées sont les `user`, les `user_profile` et les `rdvs`. Les agents n'ont pas besoin d'un traitement différencié. Lorsque ces ressources sont supprimées côté rdvs, au lieu de les supprimer, nous nullifions le rdv_solidarites_xxx_id, et faisons en sorte que l'app continue de fonctionner correctement. Le lien avec rdvs reste "cassé" pour ces ressources tant qu'aucune action n'est entreprise sur ces ressources. Dans ces cas-là, nous recréons à la volée, de manière cachée pour l'agent, la ressource côté rdv-s.

Ce qui a changé concrètement :
- suppression du bouton "Voir sur rdv-s" sur la show user parce qu'il n'est a priori plus nécessaire (et que cela permet d'éviter une surcharge de boutons à cet endroit) - vu avec l'équipe
- création d'un bouton conditionnel pour "Trouver un RDV" : si l'user est lié à un user rdvs, le bouton marche normalement ; dans le cas contraire, il s'agit d'un form qui va recréer l'usager puis rediriger vers la bonne page, comme si l'usager existait déjà
- Vérification systématique au moment d'une invit la présence d'un rdv_solidarites_user_id ; le cas échéant, l'utilisateur est recréé sur rdv-s avant de l'inviter
- quand on recrée à la volée sur rdvs, on l'ajoute automatiquement à toutes ses organisations
- la suppression par rgpd ne supprime ni l'utilisateur ni ne le retire d'aucune organisation
- pour les features de "ajouter à une organisation" ou de "ajouter un référent", l'utilisateur est recréé à la volée s'il n'existe plus côté rdv-s

N.B. : j'ai également fait une légère refonte du service `UpsertRdvSolidaritesUser` pour permettre un bon fonctionnement du système. Au passage, j'ai supprimé la logique de `existing_user` parce que j'ai l'impression que ce cas de figure ne peut pas arriver en réalité. Dans quel cas cela peut-il se produire ? Je laisse @aminedhobb me confirmer ou m'infrmer ce point. Je n'ai en tout cas pas réussi à reproduire avec les tests.

### CODE REVIEW

File: app/controllers/referent_assignations_controller.rb [hunk 1]
```diff
@@ -74,6 +77,12 @@ def remove_referent
     )
   end
 
+  def save_user
```
Comment (Holist) [inline app/controllers/referent_assignations_controller.rb:80]: Concernant le naming je me demande si ce ne serait pas plus clair de dire directement `save_to_rdvs` vu que l'on a  suprimer la liaison entre le modéle user et rdvs `save` me parait peu explicite désormais (Et ce que l'on se contente de cette méthode ou est ce que l'on change les noms des services ?)

File: app/jobs/rdv_solidarites_webhooks/process_rdv_job.rb [hunk 2]
```diff
@@ -16,13 +26,19 @@ def perform(data, meta)
...
+        upsert_rdv
         invalidate_related_invitations if created_event?
         send_outgoing_webhooks
       end
     end
 
-    private
+    def delete_or_nullify_rdv
+      if webhook_reason == "rgpd"
+        NullifyRdvSolidaritesIdJob.perform_async("Rdv", rdv&.id)
```
Comment (Holist) [inline app/jobs/rdv_solidarites_webhooks/process_rdv_job.rb:37]: 👍  Cool le petit service.

File: app/views/users/show.html.erb [hunk 3]
```diff
@@ -6,11 +6,12 @@
...
-        <button class="btn btn-blue">Voir sur RDV-Solidarités<i class="fas fa-external-link-alt icon-sm"></i></button>
-      <% end %>
-      <%= link_to rdv_solidarites_find_rdv_url(@organisation, @user), target: "_blank" do %>
-        <button class="btn btn-blue">Trouver un RDV sur RDV-Solidarités<i class="fas fa-external-link-alt icon-sm"></i></button>
+      <% if @user.rdv_solidarites_user_id? %>
+        <%= link_to rdv_solidarites_find_rdv_url(@organisation, @user), target: "_blank" do %>
+          <button class="btn btn-blue">Trouver un RDV<i class="fas fa-external-link-alt icon-sm"></i></button>
+        <% end %>
+      <% else %>
+        <%= render "create_rdvs_user_and_redirect_to_find_rdv_button", user: @user, organisation: @organisation, department: @department %>
```
Comment (Holist) [inline app/views/users/show.html.erb:14]: `create_user_in_rdvs_and_redirect_to_find_rdv_button` ? histoire de ne pas confondre avec la relation `rdvs_user` qui existe encore dans rdvs.

File: app/jobs/rdv_solidarites_webhooks/process_user_profile_job.rb [hunk 4]
```diff
@@ -4,8 +4,8 @@ def perform(data, meta)
       @data = data.deep_symbolize_keys
       @meta = meta.deep_symbolize_keys
 
-      return if @data[:user].blank?
-      return if user.blank? || organisation.blank?
+      # if webhook_reason is rgpd, the ProcessUserJob will also receive a webhook, we let him nullify the rdvs relation
+      return if @data[:user].blank? || user.blank? || organisation.blank? || webhook_reason == "rgpd"
```
Comment (Holist) [inline app/jobs/rdv_solidarites_webhooks/process_user_profile_job.rb:8]: Ce n'est pas très clair pour moi... ce serait possible d'éclairer un peu tout ces cas ? (avec des méthodes ou plus de commentaires...)

File: app/services/invitations/assign_attributes.rb [hunk 5]
```diff
@@ -22,6 +23,15 @@ def rdv_solidarites_token
       retrieve_rdv_solidarites_token.invitation_token
     end
 
+    def recreate_user_on_rdv_solidarites
+      @recreate_user_on_rdv_solidarites ||= call_service!(
+        Users::Save, # saving the user will recreate it on rdv_solidarites
```
Comment (Holist) [inline app/services/invitations/assign_attributes.rb:28]: Ce commentaire confirme qu'il y a peut être qqchose à améliorer niveau naming...

**Review by Holist — COMMENTED** [review]
Merci @qblanc !
J'ai fait quelques suggestions concernant le naming. Je laisse le soin à @aminedhobb ou @Michaelvilleneuve de faire une review plus approfondie sur le code et les améliorations possibles.

File: app/services/users/save.rb [hunk 6]
```diff
@@ -12,7 +12,10 @@ def call
         validate_user!
         save_record!(@user)
         upsert_rdv_solidarites_user
-        assign_rdv_solidarites_user_id unless @user.rdv_solidarites_user_id?
+        if @user.rdv_solidarites_user_id.nil?
+          assign_rdv_solidarites_user_id
+          assign_referents if @user.referents.present?
+        end
```
Comment (aminedhobb) [inline app/services/users/save.rb:18]: Peut-être qu'on peut extraire ça dans un service (du style `LinkToRdvSolidarites` ou `SyncWithRdvSolidarites`), comme ça on pourra appeler ce nouveau service seulement au lieu d'appeler `Users::Save` quand on voudra rattacher un user à rdv-s

File: app/services/upsert_rdv_solidarites_user.rb [hunk 7]
```diff
@@ -77,24 +60,40 @@ def user_belongs_to_org?
...
   def create_rdv_solidarites_user
     @create_rdv_solidarites_user ||= RdvSolidaritesApi::CreateUser.call(
-      user_attributes: @rdv_solidarites_user_attributes.merge(organisation_ids: [@rdv_solidarites_organisation_id]),
+      user_attributes: rdv_solidarites_user_attributes.merge(organisation_ids: rdv_solidarites_organisation_ids),
       rdv_solidarites_session: @rdv_solidarites_session
     )
   end
 
+  def rdv_solidarites_organisation_ids
+    return @user.organisations.map(&:rdv_solidarites_organisation_id) if @user.organisations.present?
```
Comment (aminedhobb) [inline app/services/upsert_rdv_solidarites_user.rb:76]: Est-ce qu'il y a des cas où cette la condition `@user.organisations.present?` renvoie `false`?

File: app/services/upsert_rdv_solidarites_user.rb [hunk 8]
```diff
@@ -19,35 +13,24 @@ def call
...
     result.errors += create_rdv_solidarites_user.errors
     fail!
   end
 
-  def existing_user
-    @existing_user ||= User.find_by(rdv_solidarites_user_id: rdv_solidarites_user_id)
-  end
-
   def assign_to_org_and_update
     create_user_profile unless user_belongs_to_org?
```
Comment (aminedhobb) [inline app/services/upsert_rdv_solidarites_user.rb:35]: est-ce qu'on a encore besoin de cette logique ? On peut pas à la place juste passer `rdv_solidarites_organisation_ids` dans les attributs de l'action `update_rdv_solidarites_user`?

File: app/views/users/_create_rdvs_user_and_redirect_to_find_rdv_button.html.erb [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<div class="d-inline-block">
+  <%= simple_form_for(user, url: "#{compute_show_path(user, organisation, department)}?create_rdvs_user_and_find_rdv=true", html: { method: :patch }, data: { turbo: false }) do |f| %>
```
Comment (aminedhobb) [inline app/views/users/_create_rdvs_user_and_redirect_to_find_rdv_button.html.erb:2]: Pour moi cette logique ne doit pas être réglée au préalable dans le front-end.
On peut garder un lien `Trouver un RDV` qui renvoie vers une action custom dans notre back-end (du style `rdv#new`) pour laquelle on pourra justement appeler le `before_action` dont pour recréer l'usager s'il n'existe plus avant de rediriger vers rdv-s.
On fait d'ailleurs qq chose de similaire dans l'action `convocations#new`.

**Review by aminedhobb — COMMENTED** [review]
Merci pour ce travail @qblanc 🙌 !!
Je pense que c'est un bon début mais qu'on peut rendre le process encore plus fluide, LMKWYT :).

File: app/services/upsert_rdv_solidarites_user.rb (hunk 8 above)
Comment (qblanc) [reply app/services/upsert_rdv_solidarites_user.rb:35]: yes bien vu

File: app/views/users/_create_rdvs_user_and_redirect_to_find_rdv_button.html.erb (hunk 9 above)
Comment (qblanc) [reply app/views/users/_create_rdvs_user_and_redirect_to_find_rdv_button.html.erb:2]: en effet c'est beaucoup plus propre

**Comment by qblanc** [comment]
j'ai profité des modifs pour supprimer un peu de code inutile - parfois à cause/grâce à des changements dans cette PR, parfois juste parce que c'était déjà du code mort (certaines méthodes du client rdv-s + le service retrieve_organisation_resource`)

**Review by aminedhobb — COMMENTED** [review]
Merci @qblanc pour les changements c'est vraiment top 🙏 🙌 !!
Je trouve que l'implémentation est super cool, c'est quand même un changement majeur et je trouve qu'il est bien amené ici 🔥 . 

J'ai fait des tests fonctionnels, globalement ça marche nickel mais il y a 2/3 edge cases supplémentaires qu'on devrait prendre en compte.

D'un point de vue métier je trouve que nos règles deviennent de plus en plus compliqués entre les règles internes à rdvi et les règles rdvs, mais je sais pas si on peut beaucoup simplifier ce process.

File: app/jobs/nullify_rdv_solidarites_id_job.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+class NullifyRdvSolidaritesIdJob < ApplicationJob
+  def perform(class_name, id)
+    resource = class_name.constantize.find_by(id: id)
+    return if resource.blank? || resource.try(:deleted?)
```
[the remark this answers is not in the document]
Comment (qblanc) [reply app/jobs/nullify_rdv_solidarites_id_job.rb:4]: je ne pense pas pouvoir faire ça parce que le modèle `Rdv`n'a pas de méthode `deleted?`, c'est pour garder cette agnosticisme du service par rapport au type de ressource passée au service que j'avais utilisé `try`

File: app/services/users/sync_with_rdv_solidarites.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+      @rdv_solidarites_session = rdv_solidarites_session
+    end
+
+    def call
+      User.transaction do
+        upsert_rdv_solidarites_user
+        if @user.rdv_solidarites_user_id.nil?
+          assign_referents if @user.referents.present?
+          assign_rdv_solidarites_user_id
+          save_record!(@user)
```
[the remark this answers is not in the document]
Comment (Holist) [reply app/services/users/sync_with_rdv_solidarites.rb:15]: Effectivement, il y aurait un travail de refacto de l'api de rdvs pour pouvoir prendre en charge cela et virer l'endpoint referent_assignation. Ca ferait pas mal de ligne en moins dans rdvs et dans rdvi.

File: app/jobs/rdv_solidarites_webhooks/process_rdv_job.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (qblanc) [reply app/jobs/rdv_solidarites_webhooks/process_rdv_job.rb:37]: yes tu as raison. Du coup j'ai modifié le `DeleteRdvJob` pour lui passer l'id plutôt que le `rdv_solidarites_rdv_id` afin d'être plus cohérent

File: app/jobs/nullify_rdv_solidarites_id_job.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (aminedhobb) [reply app/jobs/nullify_rdv_solidarites_id_job.rb:4]: Ah oui my bad j'avais pas ça en tête 👍

File: app/services/users/sync_with_rdv_solidarites.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (aminedhobb) [reply app/services/users/sync_with_rdv_solidarites.rb:15]: Oui, d'ailleurs je pense qu'on peut enlever tous les endpoints qu'on appelle plus sur RDVS en regardant toutes les méthodes du `RdvSolidaritesClient` que @qblanc  enlève ici.

Aussi il faut garder en tête que les endpoints qui permettent de créer et mettre à jour des usagers ont des autorisations intéressantes pour nous mais qui diffère de celui de l'interface web RDVS: on peut assigner n'importe quelle organisation (même celle auxquelles on appartient pas) et n'importe quel agent à un usager via l'API

[the remark this answers is not in the document]
Comment (qblanc) [reply app/services/users/sync_with_rdv_solidarites.rb]: for the record, pour ce qui est des authorisations, ce n'est pas tout à fait vrai, il y a des limites dans l'API

File: app/controllers/referent_assignations_controller.rb [hunk 12]
```diff
@@ -49,6 +51,7 @@ def agent_email
 
   def set_user
     @user = policy_scope(User).includes(:referents).find(user_id)
+    sync_user_with_rdv_solidarites(@user) if action_name == "create" && @user.rdv_solidarites_user_id.nil?
```
[the remark this answers is not in the document]
Comment (qblanc) [reply app/controllers/referent_assignations_controller.rb:54]: np, c'est mieux que ça

**Review by aminedhobb — APPROVED** [review]
Je pense qu'on y est, merci @qblanc pour la persevérence 😄 🙌 !!
Il y a un détail qu'on devrait changer je pense mais je mets quand même en "approved" parce que tout a l'air bon sinon :)

File: app/services/upsert_rdv_solidarites_user.rb [hunk 13]
```diff
@@ -61,40 +58,56 @@ def email_taken_error?
...
+                           .symbolize_keys
+                           .slice(*User::SHARED_ATTRIBUTES_WITH_RDV_SOLIDARITES)
+                           .transform_values(&:presence)
+                           .compact
+    user_attributes.delete(:email) if @user.conjoint?
+    user_attributes
+  end
+
+  def department_rdv_solidarites_organisation_ids
+    @user.organisations.where(department: @organisation.department).map(&:rdv_solidarites_organisation_id)
```
[the remark this answers is not in the document]
Comment (qblanc) [reply app/services/upsert_rdv_solidarites_user.rb:107]: ah oui tu as raison

File: app/services/rdv_solidarites_api/create_referent_assignations.rb [hunk 14]
```diff
@@ -0,0 +1,20 @@
+module RdvSolidaritesApi
+  class CreateReferentAssignations < Base
+    def initialize(rdv_solidarites_user_id:, rdv_solidarites_agent_ids:)
+      @rdv_solidarites_user_id = rdv_solidarites_user_id
+      @rdv_solidarites_agent_ids = rdv_solidarites_agent_ids
+      @rdv_solidarites_session = rdv_solidarites_session_with_shared_secret
```
[the remark this answers is not in the document]
Comment (qblanc) [reply app/services/rdv_solidarites_api/create_referent_assignations.rb:6]: yes, je l'ai rajouté à la source dans le base service

File: app/services/upsert_rdv_solidarites_user.rb [hunk 15]
```diff
@@ -1,38 +1,36 @@
...
-    @rdv_solidarites_user_id = rdv_solidarites_user_id
   end
 
   def call
     upsert_rdv_solidarites_user
-    result.rdv_solidarites_user_id = rdv_solidarites_user_id
+    return unless @user.rdv_solidarites_user_id.nil?
+
+    @user.rdv_solidarites_user_id = rdv_solidarites_user_id
+    save_record!(@user)
```
[the remark this answers is not in the document]
Comment (qblanc) [reply app/services/upsert_rdv_solidarites_user.rb:13]: Oui, et une raison importante : le service `UpsertRdvSolidaritesUser` est appelé indépendamment du service `Save` dans la méthode `sync_user_with_rdv_solidarites` de l'`ApplicationController`. Dans ce cas-là, il faut quand même persister le `rdv_solidarites_user_id`

[the remark this answers is not in the document]
Comment (aminedhobb) [reply app/services/upsert_rdv_solidarites_user.rb:13]: Ah oui ok 👍 . Pour moi le service ne devrait refléter ça dans son nom alors ce qui n'est pas le cas ici: On pourrait garder `SyncWithRdvSolidarites` qu'il y avait un moment.
